lib_test/protocol_test.ml

```diff
@@ -12,132 +12,322 @@
   let next = ref test_sequence in
-  let rec read buf = match !next with
-    | `Server x :: rest ->
-      let available = min (Cstruct.len buf) (String.length x) in
-      Cstruct.blit_from_string x 0 buf 0 available;
-      next := if available = String.length x then rest else `Server (String.sub x available (String.length x - available)) :: rest;
-      let buf = Cstruct.shift buf available in
-      if Cstruct.len buf = 0
-      then Lwt.return ()
-      else read buf
-    | `Client _ :: _ -> Lwt.fail_with "Client tried to read but it should have written"
-    | [] -> Lwt.fail_with "Client tried to read but the stream was empty" in
-  let rec write buf = match !next with
-    | `Server _ :: _ -> Lwt.fail_with "Client tried to write but it should have read"
-    | `Client x :: rest ->
-      let available = min (Cstruct.len buf) (String.length x) in
-      let written = String.sub (Cstruct.to_string buf) 0 available in
-      let expected = String.sub x 0 available in
-      OUnit.assert_equal ~msg:(Printf.sprintf "client wrote bytes: '%s' (length: %d), expected: '%s' (length: %d)" (String.escaped written) (String.length written) (String.escaped expected) (String.length expected)) expected written;
-      next := if available = String.length x then rest else `Client (String.sub x available (String.length x - available)) :: rest;
-      let buf = Cstruct.shift buf available in
-      if Cstruct.len buf = 0
-      then Lwt.return ()
-      else write buf
-    | [] -> Lwt.fail_with "Client tried to write but the stream was empty" in
+  let rec read buf =
+    (* Ignore reads and writes of length 0 and treat them as a no-op *)
+    if Cstruct.len buf = 0 then Lwt.return_unit else
+      match !next with
+      | (source, x) :: rest ->
+        if source = role then failwith "Tried to read but should have written";
+        let available = min (Cstruct.len buf) (String.length x) in
+        Cstruct.blit_from_string x 0 buf 0 available;
+        next := if available = String.length x then rest else (source, (String.sub x available (String.length x - available))) :: rest;
+        let buf = Cstruct.shift buf available in
+        if Cstruct.len buf = 0
+        then Lwt.return ()
+        else read buf
+      | [] -> Lwt.fail Failed_to_read_empty_stream in
+  let rec write buf =
+    (* Ignore reads and writes of length 0 and treat them as a no-op *)
+    if Cstruct.len buf = 0 then Lwt.return_unit else
+      match !next with
+      | (source, x) :: rest ->
+        if source <> role then failwith "Tried to write but should have read";
+        let available = min (Cstruct.len buf) (String.length x) in
+        let written = String.sub (Cstruct.to_string buf) 0 available in
+        let expected = String.sub x 0 available in
+        Alcotest.(check string)
+          (Printf.sprintf "wrote bytes: '%s' (length: %d), expected: '%s' (length: %d)" (String.escaped written) (String.length written) (String.escaped expected) (String.length expected))
```

Comment:
This line is a bit long, consider breaking it up over two lines.

Reply:
Indeed, that sounds good, I'll break it up.

Reply:
Actually we don't need this, because Alcotest has reasonable error reporting, so I'll just put a simple constant message here instead.